mat64/eigen.go

```diff
@@ -23,6 +29,90 @@ func symmetric(m *Dense) bool {
 	return true
 }
 
+// EigenSym is a type for creating and manipulating the Eigen decomposition of
+// symmetric matrices.
+type EigenSym struct {
+	vectorsComputed bool
+
+	values  []float64
+	vectors *Dense
+}
+
+// Factorize computes the Eigen decomposition of the symmetric matrix a.
+// The Eigen decomposition is defined as
+//  A = P * D * P^-1
+// where D is a diagonal matrix containing the eigenvalues of the matrix, and
+// P is a matrix of the eigenvectors of A. If the vectors input argument is
+// false, the eigenvectors are not computed.
+//
+// Factorize returns whether the decomposition succeeded. If the decomposition
+// failed, routines that require a successful factorization will panic.
+func (s *EigenSym) Factorize(a Symmetric, vectors bool) (ok bool) {
+	n := a.Symmetric()
+	sd := NewSymDense(n, nil)
+	sd.CopySym(a)
+
+	jobz := lapack.EigValueOnly
+	if vectors {
+		jobz = lapack.EigDecomp
+	}
+	w := make([]float64, n)
+	work := make([]float64, 1)
+	lapack64.Syev(jobz, sd.mat, w, work, -1)
+
+	work = make([]float64, int(work[0]))
+	ok = lapack64.Syev(jobz, sd.mat, w, work, len(work))
+	if !ok {
+		s.vectorsComputed = false
+		s.values = nil
+		s.vectors = nil
+		return false
+	}
+	s.vectorsComputed = vectors
+	s.values = w
+	s.vectors = NewDense(n, n, sd.mat.Data)
+	return true
+}
+
+// succFact returns true if the receiver contains a successful factorization.
+func (es *EigenSym) succFact() bool {
+	return len(es.values) != 0
+}
+
+// Values extracts the eigenvalues of the factorized matrix. If destination is
+// non-nil, the values are stored in-place into the receiver. In this case the
+// destination must have length n, otherwise Values will panic. If dst is
+// nil, then a new slice will be allocated of the proper length.
+//
+// Values panics if the Eigen decomposition was not successful.
+func (es *EigenSym) Values(dst []float64) []float64 {
+	if !es.succFact() {
+		panic(badFact)
+	}
+	if dst == nil {
+		dst = make([]float64, len(es.values))
+	}
+	if len(dst) != len(es.values) {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(dst, es.values)
+	return dst
+}
+
+// EigenvectorsSym extracts the eigenvectors of the factorized matrix and stores
+// them in the reciever. Each eigenvector is a column of the matrix, with the
```

Comment:
I don't like much the second part of this paragraph. How about:
```
... and stores them in the receiver whose columns are the eigenvectors corresponding to eigenvalues returned by Values.
```
or something like that?